### ISSUE TITLE
Tartube always saving description / metadata / annotations in .temp dir not in download dir

### ISSUE THREAD

**Issue by Lod3** [axcore/tartube#263]
I am writing a guide to use Tartube to archive YouTube videos. 
Unfortunately I noticed some settings I used in Tartube are not being saved? Or at least do not seem to work, maybe some of the settings are overwriting others.  
 
The goal is to create an MKV file **accompanied** by the description file and the metadata file, with the subtitles merged into the MKV file. I selected the MKV format when a merge is needed in the "Format" tab. I have checked all options in the "Write/move files" menu. 

![python3_7lpCkfc3cK](https://user-images.githubusercontent.com/14292591/108209832-0ec11780-712b-11eb-9ec7-60e04f2b8bab.png)

Under the post-processing tab I selected MKV next to the option "Encode video to another format if necessary".
And yet if I download the video, the description file is not saved nor is the metadata file. No sub-folders are being created, I just checked this again to hopefully get the description / metadata / annotations files in there. If that would work that would be good enough. 
I can see the files being saved in the correct folder during downloading, but then are removed after post-processing, I can find the files under the .temp dir, but this cannot be by design? 

What am I missing? 
**Output:**
`youtube-dl --newline -i --hls-prefer-native -o E:/UNPAK/GUIDE/%(creator)s-%(title)s-%(id)s-%(resolution)s-%(fps)s.%(ext)s --write-description --write-info-json --write-annotations --cookies E:/UNPAK/cookies.txt --write-thumbnail -f bestvideo[height<=?1440][fps<=?60]+bestaudio/best[height<=?1440] --prefer-free-formats --merge-output-format mkv --write-sub --sub-format srt --recode-video mkv --fixup detect_or_warn --download-archive E:/UNPAK/GUIDE/ytdl-archive.txt --verbose https://www.youtube.com/watch?v=F7xawqxI260
[youtube] F7xawqxI260: Downloading webpage
[info] Writing video description to: E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.description
[info] Writing video subtitles to: E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.en.vtt
[info] Writing video description metadata as JSON to: E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.info.json
[youtube] F7xawqxI260: Downloading thumbnail ...
[youtube] F7xawqxI260: Writing thumbnail to: E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.jpg
[debug] Invoking downloader on 'https://r9---sn-uxaxoxu-cg0r.googlevideo.com/videoplayback?expire=1613591303&ei=px4tYO2TKdWVgQez0YmQAg&ip=78.23.105.87&id=o-AHJi4uNSvJM8u7dWvw2LA5aFI3wS-9lBCFAQTvXBDR-x&itag=137&aitags=133%2C134%2C135%2C136%2C137%2C160%2C242%2C243%2C244%2C247%2C248%2C278&source=youtube&requiressl=yes&mh=yl&mm=31%2C29&mn=sn-uxaxoxu-cg0r%2Csn-5hneknee&ms=au%2Crdu&mv=m&mvi=9&pl=15&initcwndbps=1735000&vprv=1&mime=video%2Fmp4&ns=c_8XMh_gTqBlDewV_cRYNg0F&gir=yes&clen=46768974&dur=357.560&lmt=1612750092538358&mt=1613569499&fvip=5&keepalive=yes&c=WEB&txp=5432434&n=Faue4RAGDEl6THgZp&sparams=expire%2Cei%2Cip%2Cid%2Caitags%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgXk6okerRS2kBFynd3wYus6Jd2-wUTlA-pOpx8ibasUECIDy0H-p5p9v5OQg3kOXKh3WL_tvS61VAvcax0PY-BfRF&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgYfOPW2hwq7y18h8odpEdKohaD9SWFsBlFy63PumfrH4CIEjbBh1SbXOyDY1JFMAYiBZ5CRX6Gx5NkUdSma_RNFfg'
[download] Destination: E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.f137.mp4
[download] 100% of 44.60MiB in 00:11
[debug] Invoking downloader on 'https://r9---sn-uxaxoxu-cg0r.googlevideo.com/videoplayback?expire=1613591303&ei=px4tYO2TKdWVgQez0YmQAg&ip=78.23.105.87&id=o-AHJi4uNSvJM8u7dWvw2LA5aFI3wS-9lBCFAQTvXBDR-x&itag=251&source=youtube&requiressl=yes&mh=yl&mm=31%2C29&mn=sn-uxaxoxu-cg0r%2Csn-5hneknee&ms=au%2Crdu&mv=m&mvi=9&pl=15&initcwndbps=1735000&vprv=1&mime=audio%2Fwebm&ns=c_8XMh_gTqBlDewV_cRYNg0F&gir=yes&clen=6527063&dur=357.621&lmt=1612728133697009&mt=1613569499&fvip=5&keepalive=yes&c=WEB&txp=5432434&n=Faue4RAGDEl6THgZp&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgbm0_JaqDnrSQbZgo6paJki5kOVCjxBfGWxfhl6X6tF0CIApKwBsVni5h9VtdU1Ju2iVjwNnIZ9-0GhU1woRMZfwu&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgYfOPW2hwq7y18h8odpEdKohaD9SWFsBlFy63PumfrH4CIEjbBh1SbXOyDY1JFMAYiBZ5CRX6Gx5NkUdSma_RNFfg'
[download] Destination: E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-NA-NA.f251.webm
[download] 100% of 6.22MiB in 00:01
[ffmpeg] Merging formats into "E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.f137.mp4" -i "file:E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-NA-NA.f251.webm" -c copy -map "0:v:0" -map "1:a:0" "file:E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.temp.mkv"
Deleting original file E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.f137.mp4 (pass -k to keep)
Deleting original file E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-NA-NA.f251.webm (pass -k to keep)
[ffmpeg] Not converting video file E:/UNPAK/GUIDE/NA-How Law Enforcement Breaks into iPhones-F7xawqxI260-1920x1080-25.mkv - already is in target format mkv
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--newline', '-i', '--hls-prefer-native', '-o', 'E:/UNPAK/GUIDE/%(creator)s-%(title)s-%(id)s-%(resolution)s-%(fps)s.%(ext)s', '--write-description', '--write-info-json', '--write-annotations', '--cookies', 'E:/UNPAK/cookies.txt', '--write-thumbnail', '-f', 'bestvideo[height<=?1440][fps<=?60]+bestaudio/best[height<=?1440]', '--prefer-free-formats', '--merge-output-format', 'mkv', '--write-sub', '--sub-format', 'srt', '--recode-video', 'mkv', '--fixup', 'detect_or_warn', '--download-archive', 'E:/UNPAK/GUIDE/ytdl-archive.txt', '--verbose', 'https://www.youtube.com/watch?v=F7xawqxI260']
[debug] Encodings: locale cp1252, fs utf-8, out cp1252, pref cp1252
[debug] youtube-dl version 2021.02.10
[debug] Python version 3.8.3 (CPython) - Windows-10-10.0.19041-SP0
[debug] exe versions: ffmpeg 4.3, ffprobe 4.3, rtmpdump 2.4
[debug] Proxy map: ()
WARNING: No subtitle format found matching "srt" for language en, using vtt
WARNING: There are no annotations to write.
`

**Update:** 
I just tried this with the classic youtube-dl-gui interface tab, and the files are being downloaded and saved correctly. 
Any ideas on what to do with the Tartube settings to get it to work? 

### What operating system are you using?
Windows 10

### What version of Tartube are you using?
2.3.085

**Comment by axcore** [comment]
First thing to say is that I'm not an expert in selecting the right set of download options for every situation. youtube-dl documentation is inadequate or completely missing, as a result I have no idea what a lot of youtube-dl options actually do, or which combination of similar options are not allowed. (This is the reason why Tartube does not have comprehensive documentation of its own).

Your problem, as I understand it, is that after the video file is post-processed, the description file disappears. So let's focus on that.

Click **Edit > General download options...**
Click the **Completely reset all download options to their default values**
in the dialogue window, click the **Yes** button
Click **OK**  to close the window

Now try changing the download options, one at a time. After each change, download a new video. Keep doing that until you don't get the result you expected. Make a note of which options you change and then I'll try to reproduce it.

>I just tried this with the classic youtube-dl-gui interface tab, and the files are being downloaded and saved correctly.

The Classic Mode tab uses a separate set of download options, by default. You can change this by clicking the menu button in the top-right corner of the tab.

**Comment by Lod3** [comment]
Thanks for the fast reply

> Now try changing the download options, one at a time. After each change, download a new video. Keep doing that until you don't get the result you expected. Make a note of which options you change and then I'll try to reproduce it.

I did exactly that but all the settings that sounded relevant or related did not seem to work. But then I checked all the boxes in the menu "Keep files", now the files seem to stick around after the download finished. But let's be honest, the description of those options are not clear in my opinion. "Keep ... after Tartube shutds down", means, at least to me, that if left unchecked as is the default will remove the files after Tartube application is closed. But they do not seem to be, at first glance, to be related to explicitly saving the files like the "Write/move files" options, apparantly they are.  

Another inconsistency in the settings: 
![python3_IXp9hk2nAm](https://user-images.githubusercontent.com/14292591/108341245-c3b20d80-71d9-11eb-8ecd-528ceffe12b6.png)

There is another issue, when selected the option "Download subtitles file for these languages" it wil only download the english subtitle. So it's either all subtitles with "Download all available subtitles files" or nothing or only English. Because even when I removed "English" from the list, it will still download the "English" subtitle and none of the other languages. 
![python3_0t99eB7WXe](https://user-images.githubusercontent.com/14292591/108478879-582d7600-7295-11eb-9e4d-1398ab89891c.png)

 You can test this with this video: https://www.youtube.com/watch?v=LnJwH_PZXnM
 
 Here is my settings file:
 
`{
    "script_name": "tartube",
    "script_version": "2.3.085",
    "save_date": "18 Feb 2021",
    "save_time": "11:50:04",
    "file_type": "config",
    "custom_locale": "en_GB",
    "thumb_size_custom": "small",
    "main_win_save_size_flag": false,
    "main_win_save_slider_flag": false,
    "main_win_save_width": 1000,
    "main_win_save_height": 750,
    "main_win_videos_slider_posn": 250,
    "main_win_progress_slider_posn": 250,
    "main_win_classic_slider_posn": 250,
    "gtk_emulate_broken_flag": true,
    "toolbar_hide_flag": false,
    "toolbar_squeeze_flag": true,
    "show_tooltips_flag": true,
    "show_tooltips_extra_flag": true,
    "show_custom_icons_flag": false,
    "show_small_icons_in_index_flag": false,
    "auto_expand_video_index_flag": false,
    "full_expand_video_index_flag": false,
    "disable_dl_all_flag": false,
    "show_pretty_dates_flag": true,
    "catalogue_draw_frame_flag": true,
    "catalogue_draw_icons_flag": true,
    "drag_video_path_flag": true,
    "drag_video_source_flag": false,
    "drag_video_name_flag": false,
    "drag_thumb_path_flag": false,
    "show_status_icon_flag": true,
    "close_to_tray_flag": false,
    "progress_list_hide_flag": false,
    "results_list_reverse_flag": false,
    "system_error_show_flag": true,
    "system_warning_show_flag": true,
    "operation_error_show_flag": true,
    "operation_warning_show_flag": true,
    "system_msg_keep_totals_flag": false,
    "data_dir": "C:\\Users\\Clovis\\Videos\\Social Media Archiving\\Tartube",
    "data_dir_alt_list": [
        "C:\\Users\\Clovis\\Videos\\Social Media Archiving\\Tartube"
    ],
    "data_dir_use_first_flag": true,
    "data_dir_use_list_flag": true,
    "data_dir_add_from_list_flag": true,
    "sound_custom": "bell.mp3",
    "db_backup_mode": "single",
    "show_classic_tab_on_startup_flag": false,
    "classic_custom_dl_flag": false,
    "classic_dir_list": [
        "C:\\Users\\Clovis"
    ],
    "classic_dir_previous": null,
    "classic_pending_flag": false,
    "classic_pending_list": [],
    "classic_duplicate_remove_flag": false,
    "ytdl_bin": "youtube-dl",
    "ytdl_path_default": "youtube-dl",
    "ytdl_path": "youtube-dl",
    "ytdl_path_custom_flag": false,
    "ytdl_update_dict": {
        "ytdl_update_win_64": [
            "..\\..\\..\\mingw64\\bin\\python3.exe",
            "..\\..\\..\\mingw64\\bin\\pip3-script.py",
            "install",
            "--upgrade",
            "youtube-dl"
        ],
        "ytdl_update_pip3": [
            "pip3",
            "install",
            "--upgrade",
            "youtube-dl"
        ],
        "ytdl_update_pip": [
            "pip",
            "install",
            "--upgrade",
            "youtube-dl"
        ],
        "ytdl_update_default_path": [
            "youtube-dl",
            "-U"
        ],
        "ytdl_update_local_path": [
            "youtube-dl",
            "-U"
        ],
        "ytdl_update_custom_path": [
            "python3",
            "youtube-dl",
            "-U"
        ]
    },
    "ytdl_update_list": [
        "ytdl_update_win_64",
        "ytdl_update_pip3",
        "ytdl_update_pip",
        "ytdl_update_default_path",
        "ytdl_update_local_path",
        "ytdl_update_custom_path"
    ],
    "ytdl_update_current": "ytdl_update_win_64",
    "auto_switch_output_flag": true,
    "output_size_default": 1000,
    "output_size_apply_flag": true,
    "ytdl_update_once_flag": true,
    "ytdl_fork": null,
    "ytdl_output_system_cmd_flag": true,
    "ytdl_output_stdout_flag": true,
    "ytdl_output_ignore_json_flag": true,
    "ytdl_output_ignore_progress_flag": true,
    "ytdl_output_stderr_flag": true,
    "ytdl_output_start_empty_flag": true,
    "ytdl_output_show_summary_flag": false,
    "ytdl_write_system_cmd_flag": false,
    "ytdl_write_stdout_flag": false,
    "ytdl_write_ignore_json_flag": true,
    "ytdl_write_ignore_progress_flag": true,
    "ytdl_write_stderr_flag": false,
    "ytdl_write_verbose_flag": false,
    "ytsc_write_verbose_flag": false,
    "refresh_output_videos_flag": true,
    "refresh_output_verbose_flag": false,
    "refresh_moviepy_timeout": 10,
    "disk_space_warn_flag": true,
    "disk_space_warn_limit": 1000,
    "disk_space_stop_flag": true,
    "disk_space_stop_limit": 500,
    "custom_invidious_mirror": "invidious.site",
    "custom_dl_by_video_flag": false,
    "custom_dl_divert_mode": "default",
    "custom_dl_divert_website": "",
    "custom_dl_delay_flag": false,
    "custom_dl_delay_max": 5,
    "custom_dl_delay_min": 0,
    "dl_proxy_list": [],
    "ffmpeg_path": null,
    "avconv_path": null,
    "ffmpeg_convert_webp_flag": true,
    "ytsc_path": "C:/Users/Clovis/AppData/Local/Tartube/msys64/home/user/tartube/ytsc/youtube_stream_capture.py",
    "ytsc_priority_flag": false,
    "ytsc_wait_time": 1,
    "ytsc_restart_max": 5,
    "graph_data_type": "receive",
    "graph_plot_type": "graph",
    "graph_time_period_secs": 31536000,
    "graph_time_unit_secs": 2592000,
    "graph_ink_colour": "blue",
    "operation_limit_flag": false,
    "operation_check_limit": 3,
    "operation_download_limit": 3,
    "auto_clone_options_flag": true,
    "auto_delete_options_flag": true,
    "simple_options_flag": false,
    "enable_livestreams_flag": true,
    "livestream_max_days": 7,
    "livestream_use_colour_flag": true,
    "livestream_simple_colour_flag": false,
    "livestream_auto_notify_flag": false,
    "livestream_auto_alarm_flag": false,
    "livestream_auto_open_flag": false,
    "livestream_auto_dl_start_flag": false,
    "livestream_auto_dl_stop_flag": false,
    "scheduled_livestream_flag": true,
    "scheduled_livestream_wait_mins": 3,
    "scheduled_livestream_last_time": 0,
    "scheduled_livestream_extra_flag": true,
    "autostop_time_flag": false,
    "autostop_time_value": 1,
    "autostop_time_unit": "hours",
    "autostop_videos_flag": false,
    "autostop_videos_value": 100,
    "autostop_size_flag": false,
    "autostop_size_value": 1,
    "autostop_size_unit": "GiB",
    "operation_auto_update_flag": false,
    "operation_save_flag": true,
    "operation_sim_shortcut_flag": true,
    "operation_auto_restart_flag": false,
    "operation_auto_restart_time": 10,
    "operation_auto_restart_network_flag": false,
    "operation_auto_restart_max": 3,
    "operation_dialogue_mode": "dialogue",
    "operation_convert_mode": "channel",
    "use_module_moviepy_flag": true,
    "dialogue_copy_clipboard_flag": true,
    "dialogue_keep_open_flag": false,
    "allow_ytdl_archive_flag": true,
    "classic_ytdl_archive_flag": false,
    "apply_json_timeout_flag": true,
    "track_missing_videos_flag": false,
    "track_missing_time_flag": false,
    "track_missing_time_days": 14,
    "ignore_child_process_exit_flag": true,
    "ignore_http_404_error_flag": false,
    "ignore_data_block_error_flag": false,
    "ignore_merge_warning_flag": false,
    "ignore_missing_format_error_flag": false,
    "ignore_no_annotations_flag": true,
    "ignore_no_subtitles_flag": true,
    "ignore_yt_copyright_flag": false,
    "ignore_yt_age_restrict_flag": false,
    "ignore_yt_uploader_deleted_flag": false,
    "ignore_yt_payment_flag": false,
    "ignore_custom_msg_list": [],
    "ignore_custom_regex_flag": false,
    "num_worker_default": 2,
    "num_worker_apply_flag": true,
    "num_worker_bypass_flag": true,
    "bandwidth_default": 500,
    "bandwidth_apply_flag": false,
    "video_res_default": "720p",
    "video_res_apply_flag": false,
    "match_method": "exact_match",
    "match_first_chars": 10,
    "match_ignore_chars": 10,
    "auto_delete_flag": false,
    "auto_delete_watched_flag": false,
    "auto_delete_days": 30,
    "delete_on_shutdown_flag": false,
    "open_temp_on_desktop_flag": false,
    "complex_index_flag": false,
    "catalogue_mode": "grid_show_parent",
    "catalogue_mode_type": "grid",
    "catalogue_page_size": 50,
    "catalogue_show_filter_flag": false,
    "catalogue_sort_mode": "default",
    "catologue_use_regex_flag": false
}`

**Comment by axcore** [comment]
>But let's be honest, the description of those options are not clear in my opinion. "Keep ... after Tartube shutds down", means, at least to me, that if left unchecked as is the default will remove the files after Tartube application is closed.

Ok, I think I understand the source of the confusion. 

The first thing to say is that the simple layout of the edit window disguises some of the complexity of what is happening under the hood. Sometimes files (like thumbnails) are downloaded by youtube-dl, so they are downloaded into the same folder as the video. Sometimes they are directly downloaded by Tartube, and in this case they are stored in a temporary folder. It used to be the case that the temporary folders are deleted when Tartube shuts down. Later I realised that was a bad idea, because any temporary _videos_ are deleted before the user has had a chance to watch them. So now, in most cases, temporary folders are deleted when Tartube restarts.

 "Keep ... after Tartube shutds down" is indeed misleading, as it stands. I will have a think about how best to improve it.

>Another inconsistency in the settings:

Ok, the second line should say "Video uploader ID" or something, I will change that.

**Comment by axcore** [comment]
>There is another issue, when selected the option "Download subtitles file for these languages" it wil only download the english subtitle.

Confirmed, I will produce a fix for that ASAP.

**Comment by Lod3** [comment]
Hi, 

> Ok, I think I understand the source of the confusion.
I am glad you did, because I read my original sentence again and it was an unclear mess in itself. 

When a user ticks the options under "Write/move files"  doesn't that cover the same expectancy as when a user checks the "Keep files" options? Even if under the hood different things happen. The options under "Keep files" do not mention "Temporary files" maybe the confusion could be solved with that, or removing the option. Just a suggestion.

I look forward to the changes to the subtitles option and all the rest, great work!

**Comment by axcore** [comment]
I checked the code directly to remind myself of what is going on. Some of what I wrote above is not correct.

**File write options** are passed directly to youtube-dl (except video thumbnails; when "checking" a video, the thumbnail is downloaded directly by Tartube).

**File move options** and everything in the **Keep files** tab are Tartube actions. These actions occur as soon as Tartube detects that the video has been checked/downloaded successfully. 

If the "Keep" option is not selected, the file is moved to a temporary directory. (Therefore if the video thumbnail has been downloaded but you don't want to keep it, the thumbnail remains visible in Tartube's window until shutdown.) 

The temporary directory is definitely emptied both when Tartube shuts down, and when it restarts.

I have changed the labels from **Keep the description file after Tartube shuts down** to **Keep the description file after the download has finished**. Probably the user doesn't need to know that the file still exists in a temporary directory.
